### ISSUE TITLE
Exception occured while build help

### ISSUE THREAD

**Issue by steffenroeber** [sphinx-doc/sphinx#8186]
Following callstack can be seen. It seems it has problems with Q_DECL_CONSTEXPR:

Exception occurred:
  File "c:\program files (x86)\python38-32\lib\site-packages\sphinx\domains\cpp.py", line 6126, in _parse_type
    raise self._make_multi_error(prevErrors, header) from exTyped
sphinx.util.cfamily.DefinitionError: Error when parsing function declaration.
If the function has no return type:
  Error in declarator or parameters-and-qualifiers
  Invalid C++ declaration: Expecting "(" in parameters-and-qualifiers. [error at 133]
    template<typename SourceType, typename DestType, typename Rep = typename DestType::result_type, typename ValueType> Q_DECL_CONSTEXPR Rep convert (ValueType value)
    -------------------------------------------------------------------------------------------------------------------------------------^
If the function has a return type:
  Error in declarator or parameters-and-qualifiers
  If pointer to member declarator:
    Invalid C++ declaration: Expected '::' in pointer to member (function). [error at 137]
      template<typename SourceType, typename DestType, typename Rep = typename DestType::result_type, typename ValueType> Q_DECL_CONSTEXPR Rep convert (ValueType value)
      -----------------------------------------------------------------------------------------------------------------------------------------^
  If declarator-id:
    Invalid C++ declaration: Expecting "(" in parameters-and-qualifiers. [error at 137]
      template<typename SourceType, typename DestType, typename Rep = typename DestType::result_type, typename ValueType> Q_DECL_CONSTEXPR Rep convert (ValueType value)
      -----------------------------------------------------------------------------------------------------------------------------------------^
The full traceback has been saved in C:\Users\roebst2\AppData\Local\Temp\sphinx-err-k_qy8m52.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!

**Comment by jakobandersen** [comment]
Right, there is no preprocessor so you will need to expand the macro manually (to ``constexpr`` I guess). However, in this case I think it would work with adding ``Q_DECL_CONSTEXPR`` as a custom attribute ID to parser via the [``cpp_id_attributes``](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-cpp_id_attributes) config var.

**Comment by steffenroeber** [comment]
Ah, I see, Thank you.

**Comment by steffenroeber** [comment]
Unfortunately it still doesn't work:
I added 
```
cpp_id_attributes = ['Q_DECL_CONSTEXPR', 'Q_DECL_RELAXED_CONSTEXPR', 'CBLEXPORT', 'Q_DECL_IMPORT', 'Q_DECL_NOEXCEPT']
cpp_paren_attributes = ['Q_DECLARE_METATYPE']
```

and now I get:
```
Exception occurred:
  File "c:\program files (x86)\python38-32\lib\site-packages\sphinx\domains\cpp.py", line 6126, in _parse_type
    raise self._make_multi_error(prevErrors, header) from exTyped
sphinx.util.cfamily.DefinitionError: Error when parsing function declaration.
If the function has no return type:
  Error in declarator or parameters-and-qualifiers
  Invalid C++ declaration: Expected identifier in nested name, got keyword: template [error at 102]
    Q_DECLARE_METATYPE (c2DIntLine) Q_DECLARE_METATYPE(c2DFltLine) Q_DECLARE_METATYPE(c2DDblLine) template< typename T > Q_DECL_RELAXED_CONSTEXPR inline QTextStream &operator<<(QTextStream &stream
    ------------------------------------------------------------------------------------------------------^
If the function has a return type:
  Invalid C++ declaration: Expected identifier in nested name, got keyword: template [error at 102]
    Q_DECLARE_METATYPE (c2DIntLine) Q_DECLARE_METATYPE(c2DFltLine) Q_DECLARE_METATYPE(c2DDblLine) template< typename T > Q_DECL_RELAXED_CONSTEXPR inline QTextStream &operator<<(QTextStream &stream
    ------------------------------------------------------------------------------------------------------^
The full traceback has been saved in C:\Users\roebst2\AppData\Local\Temp\sphinx-err-99p301f6.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

**Comment by jakobandersen** [comment]
Indeed. This trick only works if the macro is in a position where an attribute could be (and we have changed the parser to accept attributes there).
I'm not familiar with Qt and the ``Q_DECLARE_METATYPE`` macro, but from https://doc.qt.io/qt-5/qmetatype.html I'm not sure if it should be in the declarations Sphinx gets.

**Comment by steffenroeber** [comment]
Unfortunately I found some more. PORTGROUP and Q_DECL_DEPRECATED_X is added to cpp_paren_attributes. 
Code looks like:
```
class cPowStation : public cStation
{
  STATIONDECL(cPowStation);
  cPowStation();

  PORTGROUP(float64) m_i1;
  cOPort<float64> m_o1;
  /**
  @sparam{exponent, Exponent of power operation., 2}
  */
  sint32 m_exponent;

  void pow(cBufferT<float64> *i_buffer);
};
```

```
If the function has no return type:
  Error in declarator or parameters-and-qualifiers
  Invalid C++ declaration: Expecting "(" in parameters-and-qualifiers. [error at 24]
    PORTGROUP (float64) m_i1
    ------------------------^
If the function has a return type:
  Error in declarator or parameters-and-qualifiers
  If pointer to member declarator:
    Invalid C++ declaration: Expected identifier in nested name. [error at 24]
      PORTGROUP (float64) m_i1
      ------------------------^
  If declarator-id:
    Invalid C++ declaration: Expected identifier in nested name. [error at 24]
      PORTGROUP (float64) m_i1
```


```
If the function has no return type:
  Error in declarator or parameters-and-qualifiers
  Invalid C++ declaration: Expected identifier in nested name. [error at 20]
    Q_DECL_DEPRECATED_X ("Use cObj::read; Will be removed in 11.10") bool readObj(cObj &obj
    --------------------^
If the function has a return type:
  Invalid C++ declaration: Expected identifier in nested name. [error at 20]
    Q_DECL_DEPRECATED_X ("Use cObj::read; Will be removed in 11.10") bool readObj(cObj &obj
    --------------------^
```

**Comment by jakobandersen** [comment]
Right, for the ``PORTGROUP`` example the attribute trick doesn't work, as it is used to "disguise" the type of the variable (I think). For ``Q_DECL_DEPRECATED_X`` I don't know, you example seems incomplete.
It would help a lot if you could provide the information requested in the bug report template, and preferably an example project for reproducing the situation.

**Comment by Amaya-aisi** [comment]
I try to run caffe(https://github.com/BVLC/caffe.git). Unfortunately, it has the same problem

`in _parse_type
    raise self._make_multi_error(prevErrors, header) from exTyped
sphinx.util.cfamily.DefinitionError: Error when parsing function declaration.
If the function has no return type:
  Error in declarator or parameters-and-qualifiers
  Invalid C++ declaration: Expected identifier in nested name, got keyword: template [error at 93]
    DEFINE_CAFFE_CPU_UNARY_FUNC (sgnbit, y[i]=static_cast< bool >((std::signbit)(x[i]))) template< typename Dtype > void caffe_cpu_scale(const int n
    ---------------------------------------------------------------------------------------------^
If the function has a return type:
  Invalid C++ declaration: Expected identifier in nested name, got keyword: template [error at 93]
    DEFINE_CAFFE_CPU_UNARY_FUNC (sgnbit, y[i]=static_cast< bool >((std::signbit)(x[i]))) template< typename Dtype > void caffe_cpu_scale(const int n
    ---------------------------------------------------------------------------------------------^
The full traceback has been saved in C:\Users\aisib\AppData\Local\Temp\sphinx-err-w7vqk7w0.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!`

**Comment by Amaya-aisi** [comment]
I try to run caffe(https://github.com/BVLC/caffe.git). Unfortunately, it has the same problem
``` in _parse_type
    raise self._make_multi_error(prevErrors, header) from exTyped
sphinx.util.cfamily.DefinitionError: Error when parsing function declaration.
If the function has no return type:
  Error in declarator or parameters-and-qualifiers
  Invalid C++ declaration: Expected identifier in nested name, got keyword: template [error at 93]
    DEFINE_CAFFE_CPU_UNARY_FUNC (sgnbit, y[i]=static_cast< bool >((std::signbit)(x[i]))) template< typename Dtype > void caffe_cpu_scale(const int n
    ---------------------------------------------------------------------------------------------^
If the function has a return type:
  Invalid C++ declaration: Expected identifier in nested name, got keyword: template [error at 93]
    DEFINE_CAFFE_CPU_UNARY_FUNC (sgnbit, y[i]=static_cast< bool >((std::signbit)(x[i]))) template< typename Dtype > void caffe_cpu_scale(const int n
    ---------------------------------------------------------------------------------------------^
The full traceback has been saved in C:\Users\aisib\AppData\Local\Temp\sphinx-err-w7vqk7w0.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!`

**Comment by jakobandersen** [comment]
As with ``PORTGROUP`` and ``Q_DECL_DEPRECATED_X``, the declaration is here obfuscated by the ``DEFINE_CAFFE_CPU_UNARY_FUNC`` macro (and the declaration seems to have been cut off early). Even wrose, this is a project-specific macro.
I don't know if you use a meta-compiler to produce the Sphinx declarations (like Breathe), but my best suggestion is to somehow give Sphinx the actual declaration you would like to document, without non-attribute macros.

**Comment by Amaya-aisi** [comment]
Thank you.
